sui_programmability/examples/deepbook/sources/critbit.move

```diff
@@ -0,0 +1,699 @@
+module deepbook::critbit {
+    use sui::tx_context::TxContext;
+    use sui::table::{Self, Table};
+
+    // <<<<<<<<<<<<<<<<<<<<<<<< Error codes <<<<<<<<<<<<<<<<<<<<<<<<
+    const ENotImplemented: u64 = 1;
+    const EExceedCapacity: u64 = 2;
+    const ETreeNotEmpty: u64 = 3;
+    const EKeyAlreadyExist: u64 = 4;
+    const ELeafNotExist: u64 = 5;
+    const EAssertFalse: u64 = 6;
+    const EIndexOutOfRange: u64 = 7;
+    const ENullParent: u64 = 8;
+    const ENullChild: u64 = 9;
+    // <<<<<<<<<<<<<<<<<<<<<<<< Error codes <<<<<<<<<<<<<<<<<<<<<<<<
+
+
+    // <<<<<<<<<<<<<<<<<<<<<<<< Constants <<<<<<<<<<<<<<<<<<<<<<<<
+    const PARTITION_INDEX: u64 = 0x8000000000000000; // 9223372036854775808
+    const MAX_U64: u64 = 0xFFFFFFFFFFFFFFFF; // 18446744073709551615
+    const MAX_CAPACITY: u64 = 0x7fffffffffffffff;
+    // <<<<<<<<<<<<<<<<<<<<<<<< Constants <<<<<<<<<<<<<<<<<<<<<<<<
+
+    struct Leaf<V> has store, drop {
+        key: u64,
+        value: V,
+        parent: u64,
+    }
+
+    struct InternalNode has store, drop {
+        mask: u64,
+        left_child: u64,
+        right_child: u64,
+        // We keep track of the parent node to make it easier to traverse the tree to retrieve the previous or the next leaf.
+        parent: u64,
+    }
+
+    // Leaves of the Critbit Tree are sorted in ascending order.
+    struct CritbitTree<V: store> has store{
+        root: u64,
+        internal_nodes: Table<u64, InternalNode>,
+        // A leaf contains orders at that price level.
+        leaves: Table<u64, Leaf<V>>,
+        min_leaf: u64,
+        max_leaf: u64,
+        next_internal_node_index: u64,
+        next_leaf_index: u64
+    }
+
+    public fun new<V: store>(ctx: &mut TxContext): CritbitTree<V> {
+        CritbitTree<V>{
+            root: PARTITION_INDEX,
+            internal_nodes: table::new(ctx),
+            leaves: table::new(ctx),
+            min_leaf: PARTITION_INDEX,
+            max_leaf: PARTITION_INDEX,
+            next_internal_node_index: 0,
+            next_leaf_index: 0
+        }
+    }
+
+    public fun size<V: store>(tree: &CritbitTree<V>): u64 {
+        table::length(&tree.leaves)
+    }
+
+    public fun is_empty<V: store>(tree: &CritbitTree<V>): bool {
+        table::is_empty(&tree.leaves)
+    }
+
+
+    // Return (key, index) the leaf with minimum value.
+    // A market buy order will start consuming liquidty from the min leaf.
+    public fun min_leaf<V: store>(_tree: &CritbitTree<V>): (u64, u64) {
```

Comment:
The Move compiler reports unused variable warnings for all vars except ones that begin with `_`, so starting a variable name with `_` + using it is typically not a good idea. `self` would be a more idiomatic name for this parameter.

Reply:
Done, we have adjusted the variable name with "_"  prefix, and unify the naming style